### PR TITLE
Content create, update & delete support through mutations

### PR DESCRIPTION
**Pull request by bdunogier**
### Create content

```
mutation CreateBlogPost {
  createBlogPost(
    languageCode: "eng-GB",
    parentLocationId: 2,
    input: {
      title: "The blog post's title",
      author: [
        {name: "John Doe", email: "johndoe@unknown.net"}
      ],
      body: {
        format: "html",
        input: "<h1>title</h1><p>A paragraph</p><p>Another <b>paragraph</b></p>"
      }
    }
  ) {
    _info { id mainLocationId }
    title
    body { html5 }
  }
}
```

### Update content

```
mutation updateBlogPost(
  $id: ID!,
  $input: BlogPostContentInput!
) {
  updateBlogPost(id: $id, input: $input) {
    _info { id currentVersionNo }
    title
  }
}
```

Update works like create, but without any mandatory field.

A versionNo argument is supported:
- if not specified, a draft is created from the current version
- if specified, it depends on the version's status:
  - if it is a draft, it is used for the update.
  - if it is an archived or published, it is used to create a draft

Then the draft gets published immediately.

### Delete content

```
mutation deleteBlogPost
{
  deleteContent(contentId: 256) {
    id
    contentId
  }
}
```

### File upload support

Example with curl (required, graphiql won't let you upload files):

```
curl -v -X POST \
  http://localhost:8000/graphql \
  -H 'Cookie: eZSESSIDWhatever=cookievalue' \
  -F 'operations={"query":"mutation CreateImage($name: String!, $alternativeText: String!, $file: FileUpload!) { createImage( parentLocationId: 51, input: { name: $name, image: {alternativeText: $alternativeText, file: $file} } ) { _info { id mainLocationId } name image { fileName alternativeText uri } } }","variables":{"file": null, "name": "2nd image upload", "alternativeText": "With generated schema"}}' \
  -F 'map={"0":["variables.file"]}' \
  -F "0"=@/Users/bdunogier/Desktop/screenshot.png
```

### Multi-file upload support

```
mutation BunchOfImages($files: [ImageUpload]!) {
  uploadFiles(locationId: 51, files: $files) {
    files {
      ... on ImageContent {
        name
        image {
          uri
        }
      }
    }
    warnings
  }
}
```

```
curl -v -X POST \
  http://localhost:8000/graphql \
  -H 'Cookie: eZSESSID98defd6ee70dfb1dea416cecdf391f58=1ei7ecn4vu4afdk9hk0k67kme2' \
  -F 'operations={"query":"mutation BunchOfImages($files: [FileUpload]!) { uploadFiles( locationId: 51, files: $files ) { files { ... on ImageContent { name image { uri } } } warnings } }","variables":{"files": [null, null, null]}}' \
  -F 'map={"0":["variables.files.0"], "1":["variables.files.1"], "2":["variables.files.2"]}' \
  -F "0"=@/path/to/image1.png \
  -F "1"=@/Users/bdunogier/Downloads/Arya.jpg \
  -F "2"=@/Users/bdunogier/Downloads/20181003_182928.jpg
```

### CODE REVIEW

File: doc/domain_schema.md [hunk 1]
```diff
@@ -3,12 +3,22 @@
 The domain content is the main GraphQL schema of eZ Platform. It is created based on the content model: 
 your types and theirs fields.
 
-Its usage requires that the GraphQL model is generated for a given repository, as a set of configuration files in the project.
-
-By default, the generated schema will list:
-- the content types groups
-- for each group the content types it contains
-- for each content type, its fields definitions, mapped to their Field Value Type
+Its usage requires that the GraphQL model is generated for a given repository,
```
Comment (bdunogier) [inline doc/domain_schema.md]: To be extracted, doesn't belong in this PR / commit.

File: doc/domain_schema.md [hunk 2]
```diff
@@ -23,13 +33,44 @@ Queries look like this:
...
 ```
 
+## Mutations
+For each content type, two mutations will be exposed: `create{ContentType}` and `update{ContentType}`:
+`createArticle`, `updateBlogPost`, ... they can be used to respectively create and update content items
+of each type. In addition, an input type is created for each of those mutations: `{ContentType}ContentCreateInput`,
+`{Article}ContentUpdateInput`, used to provide input data to the mutations.
+
+### Authentication
+The current user needs to be authorized to perform the operation. You can log in using `/login` to get a session cookie,
```
Comment (bdunogier) [inline doc/domain_schema.md]: Add an example of REST authentication. Or add a mutation to login ?

File: doc/howto/create_image_content.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+```
```
Comment (bdunogier) [inline doc/howto/create_image_content.md]: Add a few sentences describing the operation.

File: src/DependencyInjection/EzSystemsEzPlatformGraphQLExtension.php [hunk 4]
```diff
@@ -44,12 +45,20 @@ public function prepend(ContainerBuilder $container)
 
     private function getGraphQLConfig()
     {
-        $schemaFilePath = self::DOMAIN_SCHEMA_FILE;
+        $schema = [
+            'platform' => [
+                'query' => 'Platform',
+                'mutation' => 'PlatformMutation',
+                'resolver_maps' => ['BD\EzPlatformGraphQLBundle\GraphQL\Resolver\Map\UploadMap'],
```
Comment (bdunogier) [inline src/DependencyInjection/EzSystemsEzPlatformGraphQLExtension.php]: Fix namespace.

File: src/GraphQL/Mutation/UploadFiles.php [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    /**
+     * @return ContentInfo
+     */
+    private function createContent(array $mapping, UploadedFile $file, $locationId): ContentInfo
+    {
+        $contentService = $this->repository->getContentService();
+        $locationService = $this->repository->getLocationService();
+
+        $contentType = $this->repository->getContentTypeService()->loadContentTypeByIdentifier($mapping['contentTypeIdentifier']);
+        // @todo use parameter for language
```
Comment (bdunogier) [inline src/GraphQL/Mutation/UploadFiles.php]: Take care of the todo item.

File: src/GraphQL/Mutation/UploadFiles.php [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        $struct = $contentService->newContentCreateStruct($contentType, 'eng-GB');
+        $struct->setField($mapping['nameFieldIdentifier'], $file->getClientOriginalName());
+
+        // @todo The type of value that is used depends on contentFieldIdentifier's type
```
Comment (bdunogier) [inline src/GraphQL/Mutation/UploadFiles.php]: Take care of the todo item.

File: src/GraphQL/Resolver/DomainContentResolver.php [hunk 7]
```diff
@@ -198,6 +353,67 @@ public function resolveContentName(ContentInfo $contentInfo)
         return $this->repository->getContentService()->loadContentByContentInfo($contentInfo)->getName();
     }
 
+    private function getInputFieldValue($input, Repository\Values\ContentType\FieldDefinition $fieldDefinition)
+    {
+        $supportedInputTypes = ['ezstring', 'ezimage', 'eztext', 'ezrichtext', 'ezauthor'];
```
Comment (bdunogier) [inline src/GraphQL/Resolver/DomainContentResolver.php]: Refactor this to `FieldValueInputMapper` objects (note: the default case at line 386 will handle quite a few field types thanks to `fromString()`.

File: src/GraphQL/Resolver/DomainContentResolver.php [hunk 8]
```diff
@@ -198,6 +353,67 @@ public function resolveContentName(ContentInfo $contentInfo)
...
+                'alternativeText' => $fieldInput['alternativeText'] ?? '',
+                'fileName' => $file->getClientOriginalName(),
+                'inputUri' => $file->getPathname(),
+                'fileSize' => $file->getSize(),
+            ]);
+        } elseif ($fieldDefinition->fieldTypeIdentifier === 'ezrichtext') {
+            $format = $input[$fieldDefinition->identifier]['format'];
+            $input = $input[$fieldDefinition->identifier]['input'];
+
+            if ($format === 'html') {
```
Comment (bdunogier) [inline src/GraphQL/Resolver/DomainContentResolver.php]: Refactor this with `RichTextFormatConverter` objects (that handle both input and output ([EZP-30212](https://jira.ez.no/browse/EZP-30212))).

File: src/Resources/config/graphql/FieldValueInput.types.yml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+AuthorInput:
+    type: input-object
+    config:
+        fields:
+            id:
+                type: "Int"
+            name:
```
Comment (bdunogier) [inline src/Resources/config/graphql/FieldValueInput.types.yml]: Add description.

File: src/Resources/config/graphql/FieldValueInput.types.yml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+AuthorInput:
+    type: input-object
+    config:
+        fields:
+            id:
```
Comment (bdunogier) [inline src/Resources/config/graphql/FieldValueInput.types.yml]: Add description.

File: src/Resources/config/graphql/FieldValueInput.types.yml [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+                type: "Int"
+            name:
+                type: "String"
+            email:
```
Comment (bdunogier) [inline src/Resources/config/graphql/FieldValueInput.types.yml]: Add description.

File: src/Resources/config/graphql/PlatformMutation.types.yml [hunk 12]
```diff
@@ -2,44 +2,74 @@ PlatformMutation:
...
+FileUpload:
+    type: custom-scalar
+    config:
+        scalarType: '@=newObject("Overblog\\GraphQLBundle\\Upload\\Type\\GraphQLUploadType")'
+
+RichTextFieldInput:
+    type: input-object
+    config:
+        fields:
+            format:
```
Comment (bdunogier) [inline src/Resources/config/graphql/PlatformMutation.types.yml]: Make `format` an argument for `input` ? Otoh, a field may allow for arguments to the format itself.

File: src/Resources/config/graphql/PlatformMutation.types.yml [hunk 13]
```diff
@@ -2,44 +2,74 @@ PlatformMutation:
...
+                type: RichTextFieldInputFormat
+            input:
+                type: String
+
+RichTextFieldInputFormat:
+    type: enum
+    config:
+        values:
```
Comment (bdunogier) [inline src/Resources/config/graphql/PlatformMutation.types.yml]: Generate based on the registered `RichTextFormatConverter` objects.

**Comment by bdunogier** [comment]
@DominikaK I have reworked the binary files upload howto, could you have a look at it ? It's the one called `doc/howto/upload_binary_files.md`.

File: src/GraphQL/Mutation/UploadFiles.php [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+        $fieldDefinition = $contentType->getFieldDefinition($mapping['contentFieldIdentifier']);
+        switch ($fieldDefinition->fieldTypeIdentifier)
+        {
+            case 'ezimage':
```
[the remark this answers is not in the document]
Comment (bdunogier) [reply src/GraphQL/Mutation/UploadFiles.php]: Actually, it doesn't make sense to handle image asset in multi-file upload, does it @webhdx ? You would only upload actual images, while image asset is meant for "real" (editorial) content with _links_ to images (through the assets fieldtype). Am I correct ?

**Comment by bdunogier** [comment]
Everything has been rebased to have smaller commits:
1. delete content
2. content mutations schema generation + implementation
3. richtext fieldtype
4. files uploads

**Comment by crosa7** [comment]
Good morning. 
Before anything else thank you for this super useful bundle.
We are developing a project using your bundle and we now need mutations but to get them we had to change our composer json to get the "dev-master" version. Would it be possible for you to tag this so that we can "composer update " it ?
Thank you

**Comment by bdunogier** [comment]
Hello @crosa7, and thank you for your feedback. Version 1.0 should be tagged tomorrow. In the meantime, you can use `^1.0@dev` (it matches `dev-master`).

**Comment by crosa7** [comment]
Thank you very much,
Keep up the good work
